src/prefect/concurrency.py

```diff
@@ -0,0 +1,88 @@
+from typing import List, Literal, Union
+from contextlib import contextmanager
+
+from prefect import get_client
+from prefect.client.schemas.responses import MinimalConcurrencyLimitResponse
+from prefect.events import Event, RelatedResource, emit_event
+from prefect._internal.concurrency.api import create_call, from_sync
+from prefect._internal.concurrency.event_loop import get_running_loop
+
+
+async def acquire_concurrency_slots(
+    names: List[str], slots: int
+) -> List[MinimalConcurrencyLimitResponse]:
+    async with get_client() as client:
+        return await client.acquire_concurrency_slots(names=names, slots=slots)
+
+
+async def release_concurrency_slots(
+    names: List[str], slots: int
+) -> List[MinimalConcurrencyLimitResponse]:
+    async with get_client() as client:
+        return await client.release_concurrency_slots(names=names, slots=slots)
+
+
+def emit_concurrency_event(
+    phase: Union[Literal["acquired"], Literal["released"]],
+    primary_limit: MinimalConcurrencyLimitResponse,
+    related_limits: List[MinimalConcurrencyLimitResponse],
+    slots: int,
+    follows: Union[Event, None] = None,
+) -> Union[Event, None]:
+    resource = {
+        "prefect.resource.id": f"prefect.concurrency-limit.{primary_limit.id}",
+        "prefect.resource.name": primary_limit.name,
+        "slots-acquired": slots,
+        "limit": primary_limit.limit,
+    }
+
+    related = [
+        RelatedResource(
+            __root__={
+                "prefect.resource.id": f"prefect.concurrency-limit.{limit.id}",
+                "prefect.resource.role": "concurrency_limit",
+            }
+        )
+        for limit in related_limits
+    ]
+
+    event = emit_event(
+        f"prefect.concurrency-limit.{phase}",
+        resource=resource,
+        related=related,
+        follows=follows,
+    )
+    return event
+
+
+@contextmanager
+def concurrency(names: Union[str, List[str]], slots: int = 1):
```

Comment:
I think a more active name would be helpful here, something like `take_concurrency` or some other verb

Reply:
I think I agree with that. The `slots` argument here is saying how many slots you want to occupy for this code block (as opposed to saying that the block is limited to `N` concurrent slots).  Disambiguating that would be helpful.  My naive read of `with concurrency("api-calls", 5):` is a declarative way of saying "up to 5 API calls can happen at a time", but what it really means is "I'm 'bout to do 5 API calls".  Not totally sold on `take_` as the verb, but I feel ya.

Reply:
I agree with both of you, `concurrency` isn't a great name. I start up a thread in the raft room to talk about the naming, going to be easier to work it out there than in GitHub.